fix: improved error messages when deps are missing

thsig opened the pull request:
Output a more informative error message when invalid/unknown
dependencies are present in a module's configuration.

thsig commented on garden-service/src/garden.ts:
```diff
@@ -837,14 +837,18 @@ export class Garden {
-      this.moduleConfigs = await resolveTemplateStrings(this.moduleConfigs, moduleConfigContext)
 
-      await this.detectCircularDependencies()
+      this.validateDependencies()
```
I moved this above the `resolveTemplateStrings` call, since `resolveTemplateStrings` was throwing an (understandably uninformative) exception when dependencies were missing.

Is this OK?

eysi09 commented on garden-service/src/util/validate-dependencies.ts:
```diff
@@ -7,23 +7,104 @@
+  let detail = {}
+
+  if (missingDepsError) {
+    errMsg = errMsg + missingDepsError.message
```
`errMsg += missingDepsError.message` would be slightly terser.

eysi09 commented on garden-service/src/util/validate-dependencies.ts:
```diff
@@ -7,23 +7,104 @@
+  }
+
+  if (circularDepsError) {
+    errMsg = errMsg + "\n" + circularDepsError.message
```
Same as above with `+=`.

thsig commented:
In https://github.com/garden-io/garden/pull/484/commits/bb0fcf92ad4c7c951f111b9552ddbb79e4631af6, I changed the tests to more granularly test `detectMissingDependencies` and  `detectCircularDependencies` (instead of just expecting an exception during `scanModules`). Also addressed @eysi09's comments.

I didn't squash these changes right away, since it's probably easier to read through the changes addressing the PR comments this way.

If there are no other comments, I'd like to squash this into one commit before merging.

CC @edvald

edvald commented on garden-service/src/util/validate-dependencies.ts:
```diff
@@ -111,10 +111,10 @@ export function detectCircularDependencies(moduleConfigs: ModuleConfig[]): Confi
+  /**
+   * There's no need to account for test dependencies here, since any circularities there
+   * are accounted for via service dependencies.
+   */
```
Hmm, can you explain that?

thsig replied:
In short, because builds/deployments/tasks/tests can't currently depend on tests, so dependencies listed in test configs can't — by themselves — introduce circularities.

I'll add this explanation to the comment.

edvald commented on garden-service/test/src/validate-dependencies.ts:
```diff
@@ -0,0 +1,96 @@
+import { expect } from "chai"
```
This file should be under `test/src/util`, right?

thsig replied:
Ah yes, true that. Will move it.
